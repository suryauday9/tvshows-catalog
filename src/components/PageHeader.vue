<template>
  <div class="home">
    <v-app-bar app color="blue-grey darken-3" dark flat>
      <router-link :to="{ name: 'showsCatalog' }" style="text-decoration: none">
        <v-btn text large>
          <v-icon x-large>mdi-youtube-tv</v-icon>
          <v-toolbar-title class="headline mx-5 display-1 font-weight-bold">
            <span class="display-1 text-uppercase font-weight-medium">TV </span>
            <span class="display-1 text-uppercase font-weight-light">Shows</span>
          </v-toolbar-title>
        </v-btn>
      </router-link>
      <v-spacer></v-spacer>
      <v-btn id='searchBar' v-if="$route.name === 'showsCatalog'" icon @click="showSearchbar()">
        <v-icon>mdi-magnify</v-icon>
      </v-btn>
    </v-app-bar>
  </div>
</template>

<script>
export default {
  name: "PageHeader",
  data() {
    return {
      drawer: false
    };
  },
  methods: {
    showSearchbar() {
      this.$store.dispatch('getSearchbarStatus');
    }
  }
};
</script>
