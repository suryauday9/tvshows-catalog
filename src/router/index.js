import Vue from "vue";
import VueRouter from "vue-router";

Vue.use(VueRouter);

const routes = [
  {
    path: "/",
    name: 'ShowsCatalog',
    component: () => import('../views/ShowsCatalog.vue')
  },
  {
    path: '/show',
    name: 'show',
    component: () =>
      import('../views/ShowDetails.vue'),
    props: true,
  },
  {
    path: '/404',
    alias: '*',
    name: 'notFound',
    component: () =>
      import('../views/Page404.vue'),
  },
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  routes
});

export default router;
