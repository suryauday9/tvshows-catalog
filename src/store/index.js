import Vue from "vue";
import Vuex from "vuex";

import {
getAllShows,
getShowDetails,
getShowThumbnails,
getShowsBySearch
} from "../services/service";

Vue.use(Vuex);

const state = {
  shows: [],
  genres: [],
  showDetails: [],
  showThumbnails: [],
  showSearchbar: false
};

export const sortByRating = (response) => {
  return response.sort((prevValue, nextValue) =>
  prevValue.rating.average < nextValue.rating.average ? 1 : -1);
};

export const getGenres = (response) => {
  return response
  .reduce((result, show) => result.concat(show.genres), [])
  .filter((genre, index, self) => self.indexOf(genre) === index);
};

export const actions = {
  async getShowslist({ commit }) {
    getAllShows().then((response) => {
      commit('setShowslist', sortByRating(response));
      commit('setGenreslist', getGenres(response));
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem getting all shows: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  async getShowDetails({ commit }, showId) {
    getShowDetails(showId).then((response) => {
      commit('setShowDetails', response);
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem getting show details: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  async getShowThumbnails({ commit }, showId) {
    await getShowThumbnails(showId).then((response) => {
      commit('setShowThumbnails', response);
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem getting show thumbnails: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  async getShowsBySearch({ commit }, searchTerm) {
    await getShowsBySearch(searchTerm).then((response) => {
      let resultSet = response.map(result => (result.show));
      commit('setShowslist', resultSet);
      commit('setGenreslist', getGenres(resultSet));
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem in show search: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  getSearchbarStatus ({ commit }) {
    commit('setSearchbarStatus', !this.state.showSearchbar);
  }
};

export const mutations = {
  setShowslist(state, payload) {
    state.shows = payload;
  },
  setGenreslist(state, payload) {
      state.genres = payload;
  },
  setShowDetails(state, payload) {
    state.showDetails = payload;
  },
  setShowThumbnails(state, payload) {
    state.showThumbnails = payload;
  },
  setSearchbarStatus(state, payload) {
    state.showSearchbar = payload;
  }
};

export default new Vuex.Store({
  state: state,
  actions: actions,
  mutations: mutations,
  modules: {}
});
