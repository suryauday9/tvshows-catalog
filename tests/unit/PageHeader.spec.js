import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Vuetify from 'vuetify';
import VueRouter from 'vue-router';

import { initializeApp } from './initializeApp';
import PageHeader from '../../src/components/PageHeader.vue';
import { routes } from '../../src/router';

const vuetify = new Vuetify();
const localVue = initializeApp(createLocalVue());

describe('PageHeader component', () => {  
    let wrapper;
    const router = new VueRouter(routes);
    
    const store = new Vuex.Store({
        actions: {
            getSearchbarStatus: jest.fn()
        }
    });

    const mountComponent = (options) => {
        return shallowMount(PageHeader, {
            localVue,
            router,
            store,
            vuetify,
            ...options,
        })
    }

    beforeEach(() => {
        wrapper = mountComponent();
    })

    afterEach(() => {
        wrapper.destroy();
    })

    it('renders component correctly', () => {
        // see if component renders
        expect(wrapper.element).toMatchSnapshot();
    });

    it('should be a Vue instance', () => {
        expect(wrapper.isVueInstance).toBeTruthy();
    });

    it('should trigger onclick event of search button', () => {
        const headerWrapper = mount(PageHeader, {
            localVue,
            vuetify,
            router
        });
        expect(headerWrapper.find('button').trigger('click')).toStrictEqual(store.dispatch('getSearchbarStatus'));
    });

    it('showSearchbar should be instance of Function', () => {
        const actn = wrapper.vm.showSearchbar();
        expect(actn).toBeUndefined();
        expect(wrapper.vm.$nextTick()).toStrictEqual(store.dispatch('getSearchbarStatus'));
    });

});