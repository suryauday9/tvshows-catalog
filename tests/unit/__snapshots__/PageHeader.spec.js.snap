// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHeader component renders component correctly 1`] = `
<div
  class="home"
>
  <v-app-bar-stub
    app="true"
    color="blue-grey darken-3"
    dark="true"
    extensionheight="48"
    flat="true"
    src=""
    tag="header"
    value="true"
  >
    <router-link-stub
      ariacurrentvalue="page"
      event="click"
      style="text-decoration: none;"
      tag="a"
      to="[object Object]"
    >
      <v-btn-stub
        activeclass=""
        large="true"
        tag="button"
        text="true"
        type="button"
      >
        <v-icon-stub
          x-large=""
          xlarge=""
        >
          mdi-youtube-tv
        </v-icon-stub>
         
        <v-toolbar-title-stub
          class="headline mx-5 display-1 font-weight-bold"
        >
          <span
            class="display-1 text-uppercase font-weight-medium"
          >
            TV 
          </span>
           
          <span
            class="display-1 text-uppercase font-weight-light"
          >
            Shows
          </span>
        </v-toolbar-title-stub>
      </v-btn-stub>
    </router-link-stub>
     
    <v-spacer-stub />
     
    <!---->
  </v-app-bar-stub>
</div>
`;
